### ISSUE TITLE
viewport units incorrect in print

### ISSUE THREAD

**Issue by fantasai** [servo/servo#19117]
Viewport units are defined to be a percentage of the initial containing block. This seems to work fine in the case of screen media, but in print they are too big, probably because they're not subtracting out the page margin etc.

**Comment by fantasai** [comment]
Bugzilla report at https://bugzilla.mozilla.org/show_bug.cgi?id=1414600

**Comment by SimonSapin** [comment]
Servo does not yet support printing or paginated layout, so I assume you mean in Stylo?

Stylo resolves viewport-percentages relative to the `nsRect` field `mVisibleArea` of `nsPresContext` which is passed form C++ by the `Servo_StyleSet_SetDevice` function. Following up callers beyond that, it looks like the code paths are the same for Stylo and Gecko’s C++ style system.

https://searchfox.org/mozilla-central/search?q=symbol:_ZN13nsPresContext14SetVisibleAreaERK6nsRect&redirect=false shows there are a few callers of `nsPresContext::SetVisibleArea`

**Comment by fantasai** [comment]
Right, and that is not the correct thing to look at for print / print preview. The nsPresContext's mVisibleArea includes the whole page box's area afaict. But the viewport unit calculation needs to only consider the page area.
